src/accountsdb/two/Unrooted.zig

```diff
@@ -0,0 +1,175 @@
+//! A database which stores unrooted account modifications.
+//!
+//! Design notes:
+//! - This design assumes that the slots provided are never more than
+//! 4096 apart from each other.
+//!
+
+const std = @import("std");
+const sig = @import("../../sig.zig");
+
+const Atomic = std.atomic.Value;
+
+const Pubkey = sig.core.Pubkey;
+const Slot = sig.core.Slot;
+const Ancestors = sig.core.Ancestors;
+const Account = sig.core.Account;
+const AccountSharedData = sig.runtime.AccountSharedData;
+
+pub const MAX_SLOTS = 4096;
+
+const Unrooted = @This();
+
+slots: []SlotIndex,
+
+pub const SlotIndex = struct {
+    // TODO: it's possible for us to pre-populate a SlotIndex with pubkeys (and null values) in
+    // advance of replaying the slot, which would let us drop this mutex entirely.
+    lock: std.Thread.RwLock,
+    slot: Slot,
+    is_empty: Atomic(bool),
+    entries: std.ArrayHashMapUnmanaged(Pubkey, AccountSharedData, Context, true),
+
+    const Context = struct {
+        pub fn hash(_: Context, pubkey: Pubkey) u32 {
+            return @bitCast(pubkey.data[0..4].*);
```

Comment:
It's common for people to grind vanity keys to target the first few bytes, so they are not the most random.

Also there are some pre-defined vanity keys in the validator that are not on the curve. Sysvars come to mind. All of them start in base58 with "Sysvar". But then all the validator vanity keys typically end in "...11111" after whatever words they wanted to say, so taking the end or middle would not be ideal either.

We could take bytes [4..8] to avoid most of these collisions, but it still leaves one: the system and vote programs.

Reply:
Sysvars and native programs have accounts. Transactions may specify them as ordinary accounts that need to be loaded just like any other account. We have code to inject sysvars into accountsdb for every slot, since they change from slot to slot.

https://github.com/Syndica/sig/blob/85f9172b89bbb109a17a0586512ae58dbcad5c08/src/replay/update_sysvar.zig#L346

Reply:
leaving open for visibility. feel free to resolve this comment